migration: Register LP wrapped Eth USDC

NunoAlexandre opened the pull request:
### To Do

- [x] fix some compilation issues
- [x] test with `try-runtime`
- [x] add `LpEthUSDC` to the `development` runtime genesis


### Try-runtime logs


<details>
<summary> Centrifuge ✅ </summary>

NOTE: I had to comment out the `CrossChainTransferabilityMigration`'s `if VERSION.spec_version != 1020` check so that the migration would run without bumping Centrifuge spec_version.

```
λ RUST_LOG=try-runtime=debug target/release/centrifuge-chain try-runtime --runtime=target/release/wbuild/centrifuge-runtime/centrifuge_runtime.wasm --chain=centrifuge-local --detailed-log-output on-runtime-upgrade --checks live --uri=wss://fullnode.parachain.centrifuge.io:443

2023-08-24 09:49:51.987  INFO main remote-ext: since no at is provided, setting it to latest finalized head, 0xbf01b1613a3f3c77262cdfde1b7eec06c11486fdd29c35f8de5d868fff389939
2023-08-24 09:49:51.988  INFO main remote-ext: since no prefix is filtered, the data for all pallets will be downloaded
2023-08-24 09:49:52.021  INFO main remote-ext: scraping key-pairs from remote at block height 0xbf01b1613a3f3c77262cdfde1b7eec06c11486fdd29c35f8de5d868fff389939
2023-08-24 09:50:02.833  INFO main remote-ext: Querying a total of 112469 keys from prefix , splitting among 8 threads, 14059 keys per thread
2023-08-24 09:50:14.345  INFO main remote-ext: adding data for hashed prefix: , took 22s
2023-08-24 09:50:14.346  INFO main remote-ext: adding data for hashed key: 3a636f6465
2023-08-24 09:50:14.551  INFO main remote-ext: adding data for hashed key: 26aa394eea5630e07c48ae0c9558cef7f9cce9c888469bb1a0dceaa129672ef8
2023-08-24 09:50:14.578  INFO main remote-ext: adding data for hashed key: 26aa394eea5630e07c48ae0c9558cef702a5c1b19ab7a04f536c519aca4983ac
2023-08-24 09:50:14.611  INFO main remote-ext: 👩‍👦 scraping child-tree data from 240 top keys, split among 8 threads, 30 top keys per thread
2023-08-24 09:50:17.257  INFO main remote-ext: initialized state externalities with storage root 0xec0ac68d712ec0dad3e50d9cb851a213975ab521ed9b89e0b357cbe8d7c53eb5 and state_version V0
2023-08-24 09:50:17.377  INFO main try-runtime::cli: original spec: RuntimeString::Owned("centrifuge")-1019, code hash: 43afa9a6423b48c48ed093aee601a10d8e6c4ed5452e30f90e0a76ba46eb654a
2023-08-24 09:50:17.443  INFO main try-runtime::cli: new spec: RuntimeString::Owned("centrifuge")-1019, code hash: b98232c42fa298d6f9a6de69c2623680d766dae4b0bb1f53aca4890f6eaaf636
2023-08-24 09:50:17.736  INFO main runtime_common::migrations::nuke: Nuke-Loans: nuking pallet...
2023-08-24 09:50:17.736  INFO main runtime_common::migrations::nuke: Nuke-Loans: storage cleared successful
2023-08-24 09:50:17.736  INFO main runtime_common::migrations::nuke: Nuke-Loans: iteration result. backend: 1 unique: 1 loops: 1
2023-08-24 09:50:17.736  INFO main runtime_common::migrations::nuke: Nuke-InterestAccrual: nuking pallet...
2023-08-24 09:50:17.736  INFO main runtime_common::migrations::nuke: Nuke-InterestAccrual: storage cleared successful
2023-08-24 09:50:17.736  INFO main runtime_common::migrations::nuke: Nuke-InterestAccrual: iteration result. backend: 2 unique: 2 loops: 2
2023-08-24 09:50:17.737  INFO main runtime_common::migrations::asset_registry: RegisterLpEthUSDC: on_runtime_upgrade: success!
2023-08-24 09:50:17.737  INFO main runtime_common::migrations::asset_registry: RegisterLpEthUSDC: post_upgrade: the token metadata looks correct!
2023-08-24 09:50:17.737  INFO main runtime::frame-support: ⚠️ ParachainSystem declares internal migrations (which *might* execute). On-chain `StorageVersion(2)` vs current storage version `StorageVersion(2)`
2023-08-24 09:50:17.737  INFO main runtime::frame-support: ⚠️ XcmpQueue declares internal migrations (which *might* execute). On-chain `StorageVersion(2)` vs current storage version `StorageVersion(2)`
2023-08-24 09:50:17.737  INFO main runtime::frame-support: ⚠️ PolkadotXcm declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `StorageVersion(1)`
2023-08-24 09:50:17.737  INFO main runtime::frame-support: ⚠️ DmpQueue declares internal migrations (which *might* execute). On-chain `StorageVersion(1)` vs current storage version `StorageVersion(1)`
2023-08-24 09:50:17.738  INFO main runtime::frame-support: ⚠️ Ethereum declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `StorageVersion(0)`
2023-08-24 09:50:17.739 DEBUG main try-runtime::cli: proof: 0x502705db4aa834e900c8... / 35 nodes
2023-08-24 09:50:17.739 DEBUG main try-runtime::cli: proof size: 3.91 KB (4004 bytes)
2023-08-24 09:50:17.739 DEBUG main try-runtime::cli: compact proof size: 2.84 KB (2908 bytes)
2023-08-24 09:50:17.739 DEBUG main try-runtime::cli: zstd-compressed compact proof 2.65 KB (2716 bytes)
2023-08-24 09:50:17.739 DEBUG main try-runtime::cli: TryRuntime_on_runtime_upgrade executed without errors.
2023-08-24 09:50:17.783  INFO main try-runtime::cli: TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = (2200000000 ps, 0 byte), total weight = (500000000000 ps, 5242880 byte) (0.44 %, 0.00 %).
```

</details>

NunoAlexandre commented on runtime/altair/src/migrations.rs:
```diff
@@ -27,23 +27,24 @@ pub type UpgradeAltair1030 = (
-	runtime_common::migrations::nuke::Migration<crate::Loans, crate::RocksDbWeight, 1>,
-	runtime_common::migrations::nuke::Migration<crate::InterestAccrual, crate::RocksDbWeight, 0>,
+	runtime_common::migrations::nuke::Migration<crate::Loans, RocksDbWeight, 1>,
+	runtime_common::migrations::nuke::Migration<crate::InterestAccrual, RocksDbWeight, 0>,
```
Note how we do NOT add the migration to the `Algol` Upgrade set since we have already registered the token there.

lemunozm replied:
Is this comment correctly associated with those lines?

NunoAlexandre replied:
Yes and no. These changes are just about dropping the `crate::` reference prefix, but I left that note here since this block is the `Upgrade` migration set for Algol.

mustermeiszer replied:
Just for future readers:

- migration not added at all to Altair runtime

NunoAlexandre commented on runtime/common/src/migrations/asset_registry.rs:
```diff
@@ -0,0 +1,107 @@
+				AccountKey20 {
+					network: None,
+					key: vec_to_fixed_array(
+						hex::decode("a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48").unwrap(),
```
Source: https://etherscan.io/token/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48

NunoAlexandre reviewed (commented):
Some notes

mustermeiszer reviewed (commented):
I would remove the migration from `Altair` tbh. Or have we decided to bring `USDC` from Ethereum mainnet to Altair?

Otherwise, looks. good.

NunoAlexandre commented:
@mustermeiszer addressed your comment. The migration only takes place on Centrifuge but we also still register the token as part of the development runtime genesis; moved the shared objects to cfg-types/tokens for the purpose of drying those bits shared across the centrifuge runtime migration and the development genesis registration.

mustermeiszer commented on runtime/common/Cargo.toml:
```diff
@@ -8,6 +8,7 @@ homepage = "https://centrifuge.io/"
 repository = "https://github.com/centrifuge/centrifuge-chain"
 
 [dependencies]
+hex = { version = "0.4.3", default_features = false }
```
I guess this can be removed now that it lives in types.
But no blocker.

NunoAlexandre replied:
Thank you for the catch 💯

mustermeiszer reviewed (dismissed):
Thanks, looks good! Clean and neat as always!! 

We should upgrade soon, the migrations are piling up. ^^

wischli commented on runtime/centrifuge/src/migrations.rs:
```diff
@@ -169,4 +171,59 @@ mod asset_registry {
+				lp_eth_usdc_metadata(
+					LiquidityPoolsPalletIndex::get(),
+					// Ethereum mainnet chain id
+					1,
```
Nit: If we know we might need this chain ID in future migrations as well, could we make this a primitive, i.e. `ETHEREUM_MAINNET_CHAIN_ID`?

NunoAlexandre replied:
Done, thank you for the suggestion!

wischli reviewed (dismissed):
LGTM! I have two nits which can totally be ignored.
